### PR TITLE
change(mac): store partial path in UserDefaults

### LINKED ISSUE #2542: change(mac): Stop storing keyboards in Documents folder
**Describe the bug**
Currently Keyman for Mac stores keyboards under the user's Documents folder. There are other locations better suited to storing program data. This also causes a security dialog to pop unnecessarily (asking for permission to access Documents).

---

**Keyman for macOS:**
 - OS: macOS Catalina
 - Keyman version: 13.0

### PR DESCRIPTION
**Pull request by sgschantz**
Rather than store the full path of keyboard files throughout the Keyman settings in the UserDefaults, simply store the relative path from the Keyman-Keyboards directory.

With the encapsulation of the settings/UserDefaults in KMSettingsRepository, this also allowed other references to the settings to access it directly rather than gaining access via the application delegate.

This is dependent on #12106 which changed the location of the Keyman data on disk. 

Fixes #2542

----

For testing purposes, the current values of Keyman's settings in the UserDefaults can be found by executing the following command in a terminal window:

`defaults read keyman.inputmethod.Keyman`

These settings can be written to a plist file in the current directory with

`defaults read keyman.inputmethod.Keyman > ./keyman-userdefaults.plist`

These settings can be imported back into the UserDefaults as follows:

`defaults import keyman.inputmethod.Keyman ./keyman-userdefaults.plist`

Other commands are also useful for testing:
delete all Keyman settings: `defaults delete keyman.inputmethod.Keyman`
delete a single setting: `defaults delete keyman.inputmethod.Keyman KMDataModelVersion`
set a single setting: `defaults write keyman.inputmethod.Keyman KMDataModelVersion 1`
get documentation: `'defaults help'`

# User Testing

## TEST_KEYBOARDS_MOVED_TO_LIBRARY

1. Use Keyman 17 for Mac and install a couple keyboards
1. Verify that these keyboards are located in ~/Documents/Keyman-Keyboards
1. Install and run the new version of Keyman for this PR
1. Verify that the keyboards files have all been moved to ~/Library/Application 

Support/keyman.inputmethod.Keyman/Keyman-Keyboards

**TEST_ARMENIAN_OPTION_ENABLED**: 

1. Open the Stickies app on the Mac and create a new note
1. Switch to the Armenian Mnemonic keyboard 
1. Type <kbd>e</kbd><kbd>w</kbd>
1. Confirm that this produces the output: եւ
1. Type <kbd>option-v</kbd><kbd>w</kbd> (this toggles a setting within Keyman for this specific keyboard)
1. Type <kbd>e</kbd><kbd>w</kbd>
1. Confirm that this same input as earlier now produces the output: և
1. In a terminal window type the command `defaults read keyman.inputmethod.Keyman`
1. Confirm that the Armenian keyboard has an option of `option_ligature_ew` with a value of `1`. 

**TEST_ARMENIAN_OPTION_DISABLED**: 

1. With the current keyboard set to Armenian Mnemonic, and after executing TEST_ARMENIAN_OPTION_ENABLED,  type <kbd>option-v</kbd><kbd>w</kbd> to disable the option `option_ligature_ew`
1. Type <kbd>e</kbd><kbd>w</kbd>
1. Confirm that this produces the output: եւ
1. In a terminal window type the command `defaults read keyman.inputmethod.Keyman`
1. Confirm that the Armenian keyboard has an option of `option_ligature_ew` with a value of `0`. 

**TEST_SELECTED_KEYBOARD**: 

1. Install the keyboard EuroLatin (SIL) and select it as the current keyboard
1. In a terminal window type the command `defaults read keyman.inputmethod.Keyman`
1. Confirm that the property `KMSelectedKeyboardKey` is set to the value `"/sil_euro_latin/sil_euro_latin.kmx"`

### CODE REVIEW

**Comment by dinakaranr** [comment]
# Test Results
I retested this issue with the attached "Keyman 18.0.86-alpha-test" build on the macOS. Here is my observation.

* **TEST_KEYBOARDS_MOVED_TO_LIBRARY (Passed):** 
1. Installed the Keyman 17.0.328 version in the MacOS Sonoma.
2. Installed the "EuroLatin(SIL)", "GFF Amharic", "Khmer Angkor" keyboards
3. Verified that the installed keyboards are located in the ~/Documents/Keyman-Keyboards.
4. Uninstalled Keyman older version.
5. Installed the keyman 18.0.86.img keyboard.
6. Verified that the keyboards moved into the ~/Library/Application Support/keyman.inputmethod.Keyman/Keyman-Keyboards
It works well. Thanks. 


* **TEST_ARMENIAN_OPTION_ENABLED (Passed):** 
1. I have followed steps which mentioned steps 1 to 9.
2. The Armenian Mnemonic toggles keys when pressing option-v and then w
3. the և letter appears after toggles happen.
4. "option_ligature_ew = 1" is showing when executing the "defaults read keyman.inputmethod.Keyman" on the terminal.
It works well.

* **TEST_ARMENIAN_OPTION_DISABLED (Passed):** 
1. I have followed steps which mentioned steps 1 to 5.
2. The Armenian Mnemonic toggles keys when pressing option-v and then w
3. the եւ letter appears after toggles happen.
4. "option_ligature_ew = 0" is showing when executing the "defaults read keyman.inputmethod.Keyman" on the terminal.
It works well.

* **TEST_SELECTED_KEYBOARD (Passed):** 
1. Install the keyboard EuroLatin (SIL) and select it as the current keyboard.
2. KMSelectedKeyboardKey = "/sil_euro_latin/sil_euro_latin.kmx" when executing the "defaults read keyman.inputmethod.Keyman" on the terminal. 
3. If the current keyboard is changed then the path of the keyboard is changed to "KMSelectedKeyboardKey"
It works well. Thank you.

File: mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodAppDelegate.m [hunk 1]
```diff
@@ -1404,11 +1351,16 @@ - (BOOL)unzipFile:(NSString *)filePath {
...
+  os_log_debug([KMLogs keyboardLog], "unzipFile, folderName: %{public}@, keyboardFolderPath: %{public}@", folderName, keyboardFolderPath);
   [self installFontsAtPath:keyboardFolderPath];
-  for (NSString *kmxFile in [self KMXFilesAtPath:keyboardFolderPath]) {
-    os_log_debug([KMLogs keyboardLog], "Adding keyboard to list of active keyboards: %{public}@", kmxFile);
-    if (![self.activeKeyboards containsObject:kmxFile])
-      [self.activeKeyboards addObject:kmxFile];
+  
+  for (NSString *kmxFile in [self getKmxFilesAtPath:keyboardFolderPath]) {
+    NSString *partialPath = [KMDataRepository.shared buildPartialPathFrom:folderName keyboardFile:[kmxFile lastPathComponent]];
+    // TODO: encapsulate this in KMSettingsRepository, insertIfNotExists
```
Comment (mcdurdin) [inline mac/Keyman4MacIM/Keyman4MacIM/KMInputMethodAppDelegate.m:1359]: TODO: encapsulate this in KMSettingsRepository, insertIfNotExists

**Review by mcdurdin — APPROVED** [review]
LGTM

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.102-alpha](https://keyman.com/downloads/releases/alpha/18.0.102)

**Pull request merged**